src/passes/CastRefining.cpp

```diff
@@ -0,0 +1,284 @@
+      // it.
+      for (auto type : subTypes.getDepthSort()) {
+        if (!abstractTypes.count(type)) {
+          continue;
+        }
+
+        std::optional<HeapType> refinedType;
+        auto& typeSubTypes = subTypes.getStrictSubTypes(type);
+        if (typeSubTypes.size() == 1) {
+          // There is only a single possibility, so we can definitely use that
+          /// one.
+          refinedType = typeSubTypes[0];
+        } else if (!typeSubTypes.empty()) {
+          // There are multiple possibilities. However, perhaps only one of them
+          // is relevant, if nothing is ever created of the others or their
+          // subtypes.
+          for (auto subType : typeSubTypes) {
+            if (createdTypesOrSubTypes.count(subType)) {
+              if (!refinedType) {
+                // This is the first relevant thing, and hopefully will remain
+                // the only one.
+                refinedType = subType;
+              } else {
+                // We've seen more than one as relevant, so we have failed to
+                // find a singleton.
+                refinedType = std::nullopt;
+                break;
+              }
+            }
+          }
+        }
+        if (refinedType) {
+          // Propagate anything from the child, to handle chains.
+          auto iter = refinableTypes.find(*refinedType);
+          if (iter != refinableTypes.end()) {
+            *refinedType = iter->second;
+          }
+
+          refinableTypes[type] = *refinedType;
+        }
+      }
+    }
+
+    // We found optimizable types. Apply them.
+    Optimizer optimizer(*this);
+    optimizer.run(getPassRunner(), module);
+
+    // Refinalize, as RefCasts may have new types now.
+    ReFinalize().run(getPassRunner(), module);
+  }
+
+  struct Optimizer : public WalkerPass<PostWalker<Optimizer>> {
+    bool isFunctionParallel() override { return true; }
+
+    CastRefining& parent;
+
+    Optimizer(CastRefining& parent) : parent(parent) {}
+
+    std::unique_ptr<Pass> create() override {
+      return std::make_unique<Optimizer>(parent);
+    }
+
+    std::optional<HeapType> getHeapTypeIfRelevant(Type type) {
+      if (type == Type::unreachable || !type.isRef()) {
+        return std::nullopt;
+      }
+
+      // We ignore basic types, as in any non-trivial module they will always
+      // have something created of their subtypes. TODO Optimize even there
+      auto heapType = type.getHeapType();
+      if (heapType.isBasic()) {
+        return std::nullopt;
+      }
+
+      return heapType;
+    }
+
+    template<typename T> void refineCast(T* curr) {
+      auto& castType = curr->getCastType();
+      if (auto heapType = getHeapTypeIfRelevant(castType)) {
+        auto iter = parent.refinableTypes.find(*heapType);
+        if (iter != parent.refinableTypes.end()) {
+          // We can refine this cast.
+          castType = Type(iter->second, castType.getNullability());
+        }
+      }
+    }
```

Comment:
Thinking more about optimizing abstract supertypes out of the type hierarchy: Since this pass has the exact analysis we would need to do that, we could just let the global type rewriter machinery replace the cast types rather than doing it manually. 

Specifically, we would want the type rewriter to replace abstract types with their single topmost instantiated subtype or with bottom if they have no instantiated subtypes.

Reply:
TypeRewriter could do it, I guess, yeah. It would be more complicated, though, I think? We need the same analysis as here, plus TypeRewriter boilerplate.

Reply:
We could actually make this much simpler. If this pass only did the analysis and the type rewriting, then all the optimizations this pass currently implements could be handled by a subsequent `OptimizeInstructions` pass.

If I did this as a follow up, I would end up deleting all the optimization code in this pass and replacing it with type rewriting code, keeping only the analysis. Maybe that's ok if we want to verify that none of the optimization tests would regress after that change.

Reply:
I think I see what you're saying, but I'm not sure it will remain that way in the future. There are in principle situations where you can refine a cast without also refining all other casts throughout the program in an identical way, which is what rewriting at the type level would require. Does that make sense or am I missing your point?

Reply:
> There are in principle situations where you can refine a cast without also refining all other casts throughout the program in an identical way

I think doing that sort of optimization would require a dataflow analysis (possibly interprocedural) to determine what specific types can flow into a specific cast. If a type has multiple topmost instantiated children but only a single topmost instantiated child flows into a particular cast, that cast can be optimized even if other casts to the same type couldn't be optimized.

If we want to implement that, we could either factor out the analysis in this pass to share with a new pass that does the data flow analysis as well, or we could add the data flow analysis and specific cast optimization to this pass as well. Either way, the optimizations currently done by this pass would be more powerful (because of the type optimizations) and also simpler if expressed as a type replacement followed by OptimizeInstructions.

Reply:
Talking offline, I can see how the refining part of this pass makes sense to do as a type rewriting. It would also remove unneeded types we've proven are safe to remove.

Reply:
Ok, rewritten as a rewrite. This does look nicer I think!

To do this I did an NFC refactor to move the TypeMapper helper class from TypeMerging to a header.